source/extensions/common/wasm/wasm_vm.h

```diff
@@ -0,0 +1,280 @@
+      _f(WasmCallVoid<4>) _f(WasmCallVoid<5>) _f(WasmCallVoid<8>) _f(WasmCallWord<0>)              \
+          _f(WasmCallWord<1>) _f(WasmCallWord<3>)
+
+// Calls out of the WASM VM.
+// 1st arg is always a pointer to raw_context (void*).
+template <size_t N> using WasmCallbackVoid = WasmFuncType<N, void, void*, Word>*;
+template <size_t N> using WasmCallbackWord = WasmFuncType<N, Word, void*, Word>*;
+
+// Using the standard g++/clang mangling algorithm:
+// https://itanium-cxx-abi.github.io/cxx-abi/abi.html#mangling-builtin
+// Extended with W = Word
+// Z = void, j = uint32_t, l = int64_t, m = uint64_t
+using WasmCallback_WWl = Word (*)(void*, Word, int64_t);
+using WasmCallback_WWm = Word (*)(void*, Word, uint64_t);
+
+#define FOR_ALL_WASM_VM_IMPORTS(_f)                                                                \
+  _f(WasmCallbackVoid<0>) _f(WasmCallbackVoid<1>) _f(WasmCallbackVoid<2>) _f(WasmCallbackVoid<3>)  \
+      _f(WasmCallbackVoid<4>) _f(WasmCallbackWord<0>) _f(WasmCallbackWord<1>)                      \
+          _f(WasmCallbackWord<2>) _f(WasmCallbackWord<3>) _f(WasmCallbackWord<4>)                  \
+              _f(WasmCallbackWord<5>) _f(WasmCallbackWord<6>) _f(WasmCallbackWord<7>)              \
+                  _f(WasmCallbackWord<8>) _f(WasmCallbackWord<9>) _f(WasmCallback_WWl)             \
+                      _f(WasmCallback_WWm)
+
+// Wasm VM instance. Provides the low level WASM interface.
+class WasmVm : public Logger::Loggable<Logger::Id::wasm> {
+public:
+  using WasmVmPtr = std::unique_ptr<WasmVm>;
+
+  virtual ~WasmVm() = default;
+  /**
+   * Return the VM identifier.
+   * @return one of WasmVmValues from well_known_names.h e.g. "envoy.wasm.vm.null".
+   */
+  virtual absl::string_view vm() PURE;
+
+  /**
+   * Whether or not the VM implementation supports cloning.
+   * @return true if the VM is clone-able.
+   */
+  virtual bool clonable() PURE;
+
+  /**
+   * Make a thread-specific copy. This may not be supported by the underlying VM system in which
+   * case it will return nullptr and the caller will need to create a new VM from scratch.
+   * @return a clone of 'this' (e.g. for a different Worker/thread).
+   */
+  virtual WasmVmPtr clone() PURE;
+  /**
+   * Load the WASM code from a file. Return true on success.
+   * @param code the WASM binary code (or registered NullVm plugin name).
+   * @param allow_precompiled if true, allows supporting VMs (e.g. WAVM) to load the binary
+   * machine code from a user-defined section of the WASM file.
+   * @return whether or not the load was successful.
+   */
+  virtual bool load(const std::string& code, bool allow_precompiled) PURE;
+  /**
+   * Link to registered function.
+   * @param debug_name user-provided name for use in error messages.
+   * @param needs_emscripten whether emscripten support should be provided (e.g.
+   * _emscripten_memcpy_bigHandler).
+   */
+  virtual void link(absl::string_view debug_name, bool needs_emscripten) PURE;
```

Comment:
Sorry I'm still confused. If linking is a one time operation, why isn't this part of load?

Reply:
We need to load the module first so that we can query it regarding ABI support that the module requires (e.g. emscripten or WASI support).  We can then register the appropriate functions for the required ABI and do the link.  This also allows us to support different module ABI versions for backward compatibility.

Reply:
OK got it. Please just flesh out the comments and say what happens if you link twice?

Reply:
Added more comments.  I outlined the order of the calls including the fact that link should be called once.  PTAL